File: apps/objc/heatbugs/HeatbugObserverSwarm.m
// Heatbugs application. Copyright (C) 1996 Santa Fe Institute.
// This library is distributed without any warranty; without even the
// implied warranty of merchantability or fitness for a particular purpose.
// See file LICENSE for details and terms of copying.

#import "HeatbugObserverSwarm.h"
#import "HeatbugModelSwarm.h"
#import <collections.h>
#import <swarmobject.h>
#import <analysis.h>

@implementation HeatbugObserverSwarm

// createBegin: here we set up the default observation parameters.

+ createBegin: aZone
{
  HeatbugObserverSwarm * obj;
  id <ProbeMap> probeMap;
  
  // Superclass createBegin to allocate ourselves.

  obj = [super createBegin: aZone];

  // Fill in the relevant parameters (only one, in this case).

  obj->displayFrequency = 1;

  // Also, build a customized probe map. Without a probe map, the default
  // is to show all variables and messages. Here we choose to
  // customize the appearance of the probe, give a nicer interface.

  probeMap = [EmptyProbeMap createBegin: aZone];
  [probeMap setProbedClass: [self class]];
  probeMap = [probeMap createEnd];

  // Add in a bunch of variables, one per simulation parameters

  [probeMap addProbe: [probeLibrary getProbeForVariable: "displayFrequency"
				    inClass: [self class]]];

  [probeMap addProbe: [[probeLibrary getProbeForMessage: "graphBug:"
			     inClass: [self class]]
			setHideResult: 1]];

  // Now install our custom probeMap into the probeLibrary.

  [probeLibrary setProbeMap: probeMap For: [self class]];

  return obj;
}

// createEnd: create objects we know we'll need. In this case, none,
// but you might want to override this.

- createEnd
{
  return [super createEnd];
}

- _worldRasterDeath_ : caller
{
  [worldRaster drop];
  worldRaster = nil;
  return self;
}

- _unhappyGraphDeath_ : caller
{
  [unhappyGraph drop];
  unhappyGraph = nil;
  return self;
}

// Create the objects used in the display of the model. This code is
// fairly complicated because we build a fair number of widgets. It's
// also a good example of how to use the display code.

- buildObjects
{
  id modelZone;					  // zone for model.
  int i;

  [super buildObjects];
  
  // First, we create the model that we're actually observing. The
  // model is a subswarm of the observer. We also create the model in
  // its own zone, so storage is segregated.

  modelZone = [Zone create: [self getZone]];
  heatbugModelSwarm = [HeatbugModelSwarm create: modelZone];
  
  // Now create probe objects on the model and ourselves. This gives a
  // simple user interface to let the user change parameters.

  createArchivedProbeDisplay (heatbugModelSwarm);
  createArchivedProbeDisplay (self);
 
  // Instruct the control panel to wait for a button event: we halt here
  // until someone hits a control panel button so the user can get a
  // chance to fill in parameters before the simulation runs

  [controlPanel setStateStopped];

  // OK - the user has specified all the parameters for the simulation.
  // Now we're ready to start.

  // First, let the model swarm build its objects.

  [heatbugModelSwarm buildObjects];

  // Now get down to building our own display objects.

  // First, create a colormap: this is a global resource, the information
  // here is used by lots of different objects.

  colormap = [XColormap create: [self getZone]];

  // Colours [0,64) are assigned to the range Red [0, 1), for heat display.

  for (i = 0; i < 64; i++)
    [colormap setColor: i ToRed: (double)i / 63.0 Green: 0 Blue: 0];

  // Colour 64 is set to green, to display heatbugs

  [colormap setColor: 64 ToName: "green"];

  // Colour 65 is set to white, used in this case below on probed heatbug.

  [colormap setColor: 65 ToName: "white"];

  // Now go in to the heatbugs in the model and set their colours to green (64)

  [[heatbugModelSwarm getHeatbugList] forEach: M(setBugColor:) : (void *) 64];
  
  // Next, create a 2d window for display, set its size, zoom factor, title.
  
  worldRaster = [ZoomRaster createBegin: [self getZone]];
  [worldRaster setWindowGeometryRecordName : "worldRaster"];
  worldRaster = [worldRaster createEnd];
  [worldRaster enableDestroyNotification: self
               notificationMethod: @selector (_worldRasterDeath_:)];
  [worldRaster setColormap: colormap];
  [worldRaster setZoomFactor: 4];
  [worldRaster setWidth: [[heatbugModelSwarm getWorld] getSizeX]
	       Height: [[heatbugModelSwarm getWorld] getSizeY]];
  [worldRaster setWindowTitle: "Heat World"];
  [worldRaster pack];				  // draw the window.

  // Now create a Value2dDisplay: this is a special object that will
  // display arbitrary 2d value arrays on a given Raster widget.

  heatDisplay = [Value2dDisplay createBegin: [self getZone]];
  [heatDisplay setDisplayWidget: worldRaster Colormap: colormap];
  [heatDisplay setDiscrete2dToDisplay: [heatbugModelSwarm getHeat]];
  [heatDisplay setDisplayMappingM: 512 C: 0];	  // turn [0,32768) -> [0,64)
  heatDisplay = [heatDisplay createEnd];

  // And also create an Object2dDisplay: this object draws heatbugs on
  // the worldRaster widget for us, and also receives probes.

  heatbugDisplay = [Object2dDisplay createBegin: [self getZone]];
  [heatbugDisplay setDisplayWidget: worldRaster];
  [heatbugDisplay setDiscrete2dToDisplay: [heatbugModelSwarm getWorld]];
  [heatbugDisplay setObjectCollection: [heatbugModelSwarm getHeatbugList]];
  [heatbugDisplay setDisplayMessage: M(drawSelfOn:)];   // draw method
  heatbugDisplay = [heatbugDisplay createEnd];

  // Also, tell the world raster to send mouse clicks to the heatbugDisplay
  // this allows the user to right-click on the display to probe the bugs.

  [worldRaster setButton: ButtonRight Client: heatbugDisplay Message: M(makeProbeAtX:Y:)];

  // Create the graph widget to display unhappiness.

  unhappyGraph = [EZGraph createBegin: [self getZone]];
  [unhappyGraph enableDestroyNotification: self
                notificationMethod: @selector (_unhappyGraphDeath_:)];
  [unhappyGraph setWindowGeometryRecordName : "graphWindow"];
  [unhappyGraph setTitle: "Unhappiness of bugs vs. time"];
  [unhappyGraph setAxisLabelsX: "time" Y: "unhappiness"];
  unhappyGraph = [unhappyGraph createEnd];

  [unhappyGraph createAverageSequence: "unhappiness"
                         withFeedFrom: [heatbugModelSwarm getHeatbugList] 
                          andSelector: M(getUnhappiness)];

  return self;
}  

- _update_
{
  if (worldRaster)
    {
      [heatDisplay display];
      [heatbugDisplay display];
      [worldRaster drawSelf];
    }
  if (unhappyGraph)
    [unhappyGraph step];
  return self;
}

// Create the actions necessary for the simulation. This is where
// the schedule is built (but not run!)
// Here we create a display schedule - this is used to display the
// state of the world and check for user input. This schedule should
// be thought of as independent from the model - in particular, you
// will also want to run the model without any display.

- buildActions
{
  [super buildActions];
  
  // First, let our model swarm build its own schedule.

  [heatbugModelSwarm buildActions];
  
  // Create an ActionGroup for display: a bunch of things that occur in
  // a specific order, but at one step of simulation time. Some of these
  // actions could be executed in parallel, but we don't explicitly
  // notate that here.

  displayActions = [ActionGroup create: [self getZone]];

  // Schedule up the methods to draw the display of the world
  // Also schedule the update of the unhappiness graph
  //  [displayActions createActionTo: unhappinessAverager message: M(update)];

  [displayActions createActionTo: self message: M(_update_)];

  // Schedule the update of the probe displays

  [displayActions createActionTo: probeDisplayManager message: M(update)];

  // Finally, schedule an update for the whole user interface code.
  // This is crucial: without this, no graphics update and the control
  // panel will be dead. It's best to put it at the end of the display schedule

  [displayActions createActionTo: actionCache message: M(doTkEvents)];

  // And the display schedule. Note the repeat interval is set from our
  // own Swarm data structure. Display is frequently the slowest part of a
  // simulation, so redrawing less frequently can be a help.

  displaySchedule = [Schedule createBegin: [self getZone]];
  [displaySchedule setRepeatInterval: displayFrequency]; // note frequency!
  displaySchedule = [displaySchedule createEnd];
  [displaySchedule at: 0 createAction: displayActions];
  
  return self;
}  

// activateIn: - activate the schedules so they're ready to run.
// The swarmContext argument has to do with what we were activated *in*.
// Typically the ObserverSwarm is the top-level Swarm, so it's activated
// in "nil". But other Swarms and Schedules and such will be activated
// inside of us.

- activateIn:  swarmContext
{
  // First, activate ourselves (just pass along the context).

  [super activateIn: swarmContext];

  // Activate the model swarm in ourselves. The model swarm is a
  // subswarm of the observer swarm.

  [heatbugModelSwarm activateIn: self];

  // Now activate our schedule in ourselves. This arranges for the
  // execution of the schedule we built.

  [displaySchedule activateIn: self];
  
  // Activate returns the swarm activity - the thing that's ready to run.

  return [self getActivity];
}

// You could override the "go" method here if you want something special
// to happen when the model and observer actually start running. But
// the default GUISwarm go is probably good enough.

- graphBug: aBug
{
  if (unhappyGraph)
    [unhappyGraph createSequence: "Bug" 
                  withFeedFrom: aBug 
                  andSelector: M(getUnhappiness)];
  return self;
}

@end

File: apps/java/jheatbugs/HeatbugObserverSwarm.java
// Java Heatbugs application. Copyright © 1999-2000 Swarm Development Group.
// This library is distributed without any warranty; without even the
// implied warranty of merchantability or fitness for a particular
// purpose.  See file COPYING for details and terms of copying.

import java.util.ArrayList;

import swarm.Globals;
import swarm.Selector;
import swarm.defobj.Zone;

import swarm.activity.Activity;
import swarm.activity.ActionGroup;
import swarm.activity.ActionGroupImpl;
import swarm.activity.Schedule;
import swarm.activity.ScheduleImpl;

import swarm.objectbase.Swarm;
import swarm.objectbase.VarProbe;
import swarm.objectbase.MessageProbe;
import swarm.objectbase.EmptyProbeMapImpl;

import swarm.gui.Colormap;
import swarm.gui.ColormapImpl;
import swarm.gui.ZoomRaster;
import swarm.gui.ZoomRasterImpl;

import swarm.analysis.EZGraph;
import swarm.analysis.EZGraphImpl;

import swarm.simtoolsgui.GUISwarm;
import swarm.simtoolsgui.GUISwarmImpl;

import swarm.space.Value2dDisplay;
import swarm.space.Value2dDisplayImpl;
import swarm.space.Object2dDisplay;
import swarm.space.Object2dDisplayImpl;

/**
This class implements the GUI-mode observer of the Heatbug model 
defined in HeatbugModelSwarm.java.

<p>
See HeatbugModelSwarm for an overview of the heatbugs application.

*/
public class HeatbugObserverSwarm extends GUISwarmImpl
{

// This defines the number of steps after which we display a snapshot of the 
// simulation; we could speed up the simulation by displaying less frequently:
public int displayFrequency = 1;

// This defines the timing of the Swarm's Actions:  
private Schedule _displaySchedule;

// This is the model we're observing:
private HeatbugModelSwarm _heatbugModelSwarm;
    public HeatbugModelSwarm getHeatbugModelSwarm ()
    { return _heatbugModelSwarm; }

// This is the index to the palette we will use to paint heat and Heatbugs:
private Colormap _colormap;

// This is the 2-dimensional display we will use to paint both heat
// and Heatbugs:
private ZoomRaster _worldRaster;

// This is the time-series graph we will use to display average
// Heatbug unhappiness:
private EZGraph _unhappyGraph;

// This is the 2-dimensional graph of the heat; we will display
// it on the ZoomRaster:
private Value2dDisplay _heatDisplay;

// This is the 2-dimensional graph of the Heatbugs; we will display
// it on the ZoomRaster, layered over the _heatDisplay:
private Object2dDisplay _heatbugDisplay;

public HeatbugObserverSwarm (Zone aZone)
{
    super(aZone);

    // Create the model that this observer will observe:
    _heatbugModelSwarm = new HeatbugModelSwarm (getZone ());

    // Create a data structure to hold the Probes:
    EmptyProbeMapImpl heatbugObserverProbeMap = new EmptyProbeMapImpl 
     (aZone, getClass ());

    // Create Probes for some variables and methods (see HeatbugModelSwarm.java
    // for an explanation of Probes, ProbeMaps, and ProbeDisplays):
    heatbugObserverProbeMap.addProbe (probeVariable ("displayFrequency"));
    heatbugObserverProbeMap.addProbe (probeMessage ("graphBug:"));

    Globals.env.probeLibrary.setProbeMap$For
     (heatbugObserverProbeMap, getClass ());

    System.out.println 
     (
"\n" +
"In each field you change in the probe display, press Enter.\n" +
"\n" +
"For method invocations, enter an appropriate value in each argument\n" +
"textbox after the method button, then click the button.\n"
     );

} /// constructor

/**
This method activates the schedules so they're ready to run.

@param swarmContext (in)
    the larger context within which this Swarm is activated; an observer swarm 
    is usually the top-level Swarm, so the context is usually null; for
    sub-Swarms such as _heatbugModelSwarm, this HeatbugObserverSwarm will be 
    the swarmContext 
*/
public Activity activateIn (Swarm swarmContext)
{
    super.activateIn (swarmContext);
    _heatbugModelSwarm.activateIn (this);
    _displaySchedule.activateIn (this);
    return getActivity();
}

/**
This method schedules the actions of this GUI observer Swarm. 

<p>
This Swarm contains the Schedules, ActionGroups, and Actions depicted in the 
following diagram. 

<xmp>
Swarm
this
|
|
|
Schedule
_displaySchedule
|
+-----------------------------------+
|                                   |
ActionGroup                         ActionGroup
updateActions                       tkActions
|                                   |
+-------------+                     |
|             |                     |
Action        Action                Action
this          probeDisplayManager   getActionCache()   
._update_()   .update()             doTkEvents()
</xmp>

<p>
See the documentation in HeatbugModelSwarm.buildActions() for an explanation
of Schedules, ActionGroups, and Actions.

*/
public Object buildActions ()
{
    super.buildActions();

    // Let the model Swarm build its own schedule:
    _heatbugModelSwarm.buildActions();

    ActionGroup updateActions = new ActionGroupImpl (getZone());
    ActionGroup tkActions = new ActionGroupImpl (getZone());

    // Define the first Action of ActionGroup updateActions:
    try
    {
    updateActions.createActionTo$message
     (this, new Selector (getClass (), "_update_", false));

    // Define the second Action of ActionGroup updateActions:
    updateActions.createActionTo$message
     (Globals.env.probeDisplayManager,
      new Selector (Globals.env.probeDisplayManager.getClass (), "update", true)
     );

    // Define the sole Action of ActionGroup tkActions:
    tkActions.createActionTo$message
     (getActionCache (),
      new Selector (getActionCache ().getClass (), "doTkEvents", true)
     );
    } catch (Exception e)
    {
        System.err.println ("Exception in setting up tkActions : "
         + e.getMessage ());
    }

    // Define the Schedule:
    _displaySchedule = new ScheduleImpl (getZone (), displayFrequency);
      // ... The repeat interval is displayFrequency, so the schedule will
      // begin once every displayFrequency steps of the simulation.
    // Insert the updateActions ActionGroup into the Schedule:
    _displaySchedule.at$createAction 
     (0, 
      // ... Execute the ActionGroup at step 0 relative to the beginning of 
      // the schedule.
      updateActions
     );
    // Insert the tkActions ActionGroup into the Schedule:
    _displaySchedule.at$createAction 
     (0, 
      // ... Execute the ActionGroup at step 0 relative to the beginning of 
      // the schedule.
      tkActions
     );

    return this;
} /// buildActions()

/**
This method creates the plots and graphs that present the 
results of the simulation. It delegates the building of the Heatbug model 
to HeatbugModelSwarm.buildObjects().

*/
public Object buildObjects ()
{
    super.buildObjects ();

    // Create probe objects on the model and on this observer, to provide
    // GUI channels for reading and writing parameters:
    Globals.env.createArchivedProbeDisplay
     (_heatbugModelSwarm, "_heatbugModelSwarm");
    Globals.env.createArchivedProbeDisplay (this, "heatbugObserverSwarm");

    // Wait here until the user clicks Start or Next after optionally changing 
    // parameters:
    getControlPanel ().setStateStopped ();

    _heatbugModelSwarm.buildObjects ();

    // Create a Colormap for displaying Heatbugs and heat:
    _colormap = new ColormapImpl (getZone ());

    // Assign colors [ 0.. 63] to shades of red, for heat display;
    // assign colors [64..127] to shades of yellow-green, for Heatbug display:
    for (double i = 0; i < 64; i++)
    {
        _colormap.setColor$ToRed$Green$Blue ((byte) i,        i / 63, 0, 0);
        _colormap.setColor$ToRed$Green$Blue ((byte) (64 + i), i / 63, 1, 0);
    }

    // Set the colors of the heatbugs from yellow through green (the higher
    // the ideal temperature, the more the yellow):
    double tempRange 
     = _heatbugModelSwarm.maxIdealTemp - _heatbugModelSwarm.minIdealTemp;
    ArrayList heatbugList = _heatbugModelSwarm.getHeatbugList ();
    for (int i = 0; i < heatbugList.size (); i++)
    {
        Heatbug bug = (Heatbug) heatbugList.get (i);
        bug.setColorIndex 
         ((byte) 
          (64 + 63 * 
           (bug.getIdealTemperature () - _heatbugModelSwarm.minIdealTemp) 
           / tempRange
          )
         );
    }

    // Create another window for display, and set its attributes:
    _worldRaster = new ZoomRasterImpl (getZone (), "_worldRaster");
    try
    {
    _worldRaster.enableDestroyNotification$notificationMethod
     (this, new Selector (getClass (), "_worldRasterDeath_", false));
    } catch (Exception e)
    {
        System.err.println ("Exception _worldRasterDeath_: " + e.getMessage ());
    }
    _worldRaster.setColormap (_colormap);
    _worldRaster.setZoomFactor (4);
    _worldRaster.setWidth$Height
     ((_heatbugModelSwarm.getWorld ()).getSizeX (),
      (_heatbugModelSwarm.getWorld ()).getSizeY ()
     );
    _worldRaster.setWindowTitle ("Heat World");
    _worldRaster.pack();                  // draw the window

    // Create a Value2dDisplay, to display the HeatSpace on the ZoomRaster:
    _heatDisplay = new Value2dDisplayImpl
     (getZone (), _worldRaster, _colormap, _heatbugModelSwarm.getHeatSpace ());

    _heatDisplay.setDisplayMappingM$C (512, 0); // map [0..32767] to [0,63]

    // The Heatbug positional data is in the Grid2d, which we can obtain from 
    // getWorld(). The display widget is the ZoomRaster _worldRaster. An 
    // Object2dDisplay knows how to draw such data on such a raster: 
    try
    {
    _heatbugDisplay = new Object2dDisplayImpl
     (getZone (),
      _worldRaster,
      _heatbugModelSwarm.getWorld (),
      new Selector (Class.forName ("Heatbug"), "drawSelfOn", false)
     );
    } catch (Exception e)
    {
        System.err.println ("Exception drawSelfOn: " + e.getMessage ());
    }

    // The Grid2d knows what Heatbugs are on it, and _heatbugDisplay has it, so
    // _heatbugDisplay could draw it without any more help from us. But it has 
    // getSizeX () times getSizeY () cells. If we give it the Heatbug list,
    // which has only numBugs elements, it can draw the Heatbugs more 
    // efficiently: 
    _heatbugDisplay.setObjectCollection
     (_heatbugModelSwarm.getHeatbugList ());

    // Tell the world raster to send mouse clicks to the
    // _heatbugDisplay. This will allow the user to right-click on the
    // display to probe the bugs:
    try
    {
    _worldRaster.setButton$Client$Message
     (3,
      _heatbugDisplay,
      new Selector (_heatbugDisplay.getClass (), "makeProbeAtX$Y", true)
     );
    } catch (Exception e)
    {
        System.err.println ("Exception makeProbeAtX$Y: " + e.getMessage ());
    }

    // Create the graph widget to display unhappiness:
    _unhappyGraph = new EZGraphImpl
     (getZone (),
      "Unhappiness of bugs vs. time",
      "time", 
      "unhappiness",
      "_unhappyGraph"
     );

    // Todo: deal with this now-commented-out code:
    // Globals.env.setWindowGeometryRecordName (_unhappyGraph, "_unhappyGraph");

    // Assign the method to be used for destroying _unhappyGraph:
    try
    {
    _unhappyGraph.enableDestroyNotification$notificationMethod
     (this,
      new Selector (getClass (), "_unhappyGraphDeath_", false)
     );
    } catch (Exception e)
    {
        System.err.println
         ("Exception _unhappyGraphDeath_: " + e.getMessage ());
    }

    // Create the mechanism for computing the average heatbug unhappiness:
    try
    {
    _unhappyGraph.createAverageSequence$withFeedFrom$andSelector
     ("unhappiness", 
      _heatbugModelSwarm.getHeatbugList (),
      new Selector (Class.forName ("Heatbug"), "getUnhappiness", false)
     );
    } catch (Exception e)
    {
        System.err.println ("Exception getUnhappiness: " + e.getMessage ());
    }
    return this;
} /// buildObjects()

public void drop ()
{
    if (_unhappyGraph != null)
        _unhappyGraph.disableDestroyNotification ();
    if (_worldRaster != null)
        _worldRaster.disableDestroyNotification ();
    super.drop ();
}

public Object graphBug (Heatbug aBug)
{
    if (_unhappyGraph != null)
    try
    {
    _unhappyGraph.createSequence$withFeedFrom$andSelector
     ("Bug", 
      aBug,
      new Selector (aBug.getClass (), "getUnhappiness", false)
     );
    } catch (Exception e)
    {
        System.err.println ("Exception graphBug: " + e.getMessage());
    }
    return this;
}

public Object _unhappyGraphDeath_ (Object caller)
{
    _unhappyGraph.drop ();
    _unhappyGraph = null;
    return this;
}

/**
This callback method defines what the observer does whenever the Schedule 
triggers it. 

*/
public Object _update_ ()
{
    if (_worldRaster != null)
    {
        _heatDisplay.display ();
        _heatbugDisplay.display ();
        _worldRaster.drawSelf ();
    }

    if (_unhappyGraph != null)
        _unhappyGraph.step ();
    return this;
}

public Object _worldRasterDeath_ (Object caller)
{
    _worldRaster.drop ();
    _worldRaster = null;
    return this;
}

private VarProbe probeVariable (String name) 
{
    return Globals.env.probeLibrary.getProbeForVariable$inClass
     (name, HeatbugObserverSwarm.this.getClass ());
}

private MessageProbe probeMessage (String name) {
    return Globals.env.probeLibrary.getProbeForMessage$inClass
     (name, HeatbugObserverSwarm.this.getClass ());
}

} /// class HeatbugObserverSwarm
